File: src/systems/portal/bff/src/routes/ui.ts
/**
 * BFF `/:svc/ui/*` routes — serve a registered service's embedded mini-portal.
 *
 * Unlike the JSON `/api` passthrough (proxyToUpstream forces `application/json`
 * and re-serializes the body), a mini-portal serves HTML/JS/CSS/font/image
 * assets, so this streams the upstream response verbatim: the upstream
 * `Content-Type` and the raw, binary-safe body. The bearer `Authorization`
 * header is forwarded exactly as for `/api`, so the shell loads the iframe
 * same-origin and never hands a token to a cross-origin frame. Conductor only
 * routes these paths when the service registers a (public) `/ui` endpoint in its
 * manifest, so this is inert for services that ship no UI.
 */
import { Router } from 'express';
import { SpanStatusCode } from '@opentelemetry/api';
import { CONDUCTOR_URL } from '../config.js';
import { logger } from '../observability/logger.js';
import { tracer } from '../observability/tracing.js';
import { upstreamRequestDuration } from '../observability/metrics.js';

export const uiRoutes = Router();

// A service name is a simple slug; the strict match also blocks any path-traversal
// attempt from smuggling through the :svc segment.
const SVC_RE = /^[a-zA-Z0-9_-]+$/;

uiRoutes.all(['/:svc/ui', '/:svc/ui/*'], async (req, res) => {
  const svc = (req.params as Record<string, string>).svc;
  if (!SVC_RE.test(svc)) {
    res.status(400).json({ error: 'invalid service name' });
    return;
  }

  // req.url is the bare conductor path (the /api mount prefix is already stripped),
  // preserving the /ui sub-path and any query string the mini-SPA needs for assets.
  const url = `${CONDUCTOR_URL}${req.url}`;
  const headers: Record<string, string> = {};
  const auth = req.headers['authorization'];
  if (typeof auth === 'string') headers['Authorization'] = auth;

  const span = tracer.startSpan(`ui ${req.method}`);
  const endTimer = upstreamRequestDuration.startTimer({ method: req.method, route: 'ui' });

  try {
    const upstream = await fetch(url, { method: req.method, headers });
    span.setAttribute('http.status_code', upstream.status);
    endTimer();

    res.status(upstream.status);
    // Pass the upstream content type through verbatim (HTML/JS/CSS/font/image) —
    // unlike the JSON proxy — and stream the body as raw bytes to stay binary-safe.
    const ct = upstream.headers.get('content-type');
    if (ct) res.setHeader('Content-Type', ct);
    const cc = upstream.headers.get('cache-control');
    if (cc) res.setHeader('Cache-Control', cc);
    res.send(Buffer.from(await upstream.arrayBuffer()));
  } catch (err) {
    span.recordException(err as Error);
    span.setStatus({ code: SpanStatusCode.ERROR });
    endTimer();
    logger.error({ err, url }, 'upstream ui fetch failed');
    res.status(502).json({ error: 'upstream unavailable' });
  } finally {
    span.end();
  }
});

File: src/systems/portal-bff/ui.go
package main

import (
	"io"
	"log/slog"
	"net/http"
	"regexp"
	"strings"
)

// A service name is a simple slug; the strict match also blocks any path-traversal
// attempt from smuggling through the service segment.
var svcRe = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)

// isServiceUIPath reports whether an /api-stripped path targets a service's
// embedded mini-portal, i.e. /{svc}/ui or /{svc}/ui/...
func isServiceUIPath(apiPath string) bool {
	trimmed := strings.TrimPrefix(apiPath, "/")
	parts := strings.SplitN(trimmed, "/", 3)
	return len(parts) >= 2 && parts[1] == "ui" && svcRe.MatchString(parts[0])
}

// handleServiceUI serves a registered service's embedded mini-portal. Unlike the
// JSON /api passthrough (which forces application/json and re-serializes the
// body), a mini-portal serves HTML/JS/CSS/font/image assets, so this streams the
// upstream response verbatim: the upstream Content-Type and the raw, binary-safe
// body. The bearer Authorization header is forwarded exactly as for /api, so the
// shell loads the iframe same-origin and never hands a token to a cross-origin
// frame.
func handleServiceUI(w http.ResponseWriter, r *http.Request) {
	// apiPath is the conductor path with the /api mount prefix stripped, e.g.
	// /blueprints/ui/asset.js — its first segment is the service name.
	apiPath := strings.TrimPrefix(r.URL.Path, "/api")
	svc := strings.SplitN(strings.TrimPrefix(apiPath, "/"), "/", 2)[0]
	if !svcRe.MatchString(svc) {
		writeJSONError(w, http.StatusBadRequest, "invalid service name")
		return
	}

	target := conductorURL + apiPath
	if r.URL.RawQuery != "" {
		target += "?" + r.URL.RawQuery
	}
	req, _ := http.NewRequestWithContext(r.Context(), r.Method, target, nil)
	if auth := r.Header.Get("Authorization"); auth != "" {
		req.Header.Set("Authorization", auth)
	}

	resp, err := httpClient.Do(req)
	if err != nil {
		slog.ErrorContext(r.Context(), "upstream ui fetch failed", "error", err, "url", target)
		writeJSONError(w, http.StatusBadGateway, "upstream unavailable")
		return
	}
	defer resp.Body.Close()

	// Pass the upstream content type through verbatim (HTML/JS/CSS/font/image) —
	// unlike the JSON proxy — and stream the body as raw bytes to stay binary-safe.
	if ct := resp.Header.Get("Content-Type"); ct != "" {
		w.Header().Set("Content-Type", ct)
	}
	if cc := resp.Header.Get("Cache-Control"); cc != "" {
		w.Header().Set("Cache-Control", cc)
	}
	w.WriteHeader(resp.StatusCode)
	_, _ = io.Copy(w, resp.Body)
}
